### PR TITLE
Add ChatGPT experiment

### PR DESCRIPTION
**Pull request by superb-owl**
This adds an experiment that asks ChatGPT to rate posts on a variety of dimensions. I ran for `curious-frame` (screenshot). Nothing super interesting so far.

Data is very noisy and had to be smoothed aggressively.

Screenshot 2023-03-26 at 3 08 38 PM

---

Only added two things outside the `experiments` folder, happy to revert:
* Added data-download instructions to README (lmk if I got it wrong)
* added poetry.lock (I think that's typically checked in?)

A few questions came up:
* I alway have problems in python where it can't load local libraries, e.g. when running `python 1_crawl_messages.py`. To get it working, I have to add `sys.path.append("../")`. Any ideas why?
* It looks like author names have been anonymized. Any way to convert? (e.g. if I see an author on the site that I want to analyze here)
* It seems like message content might be missing newlines--known issue?
* It would be nice to get _just_ the main practice log for an author, rather than all their posts. I realize that's probably not super straightforward, but one heuristic might be "find the thread with the most posts by $AUTHOR, and filter for only their top-level posts." I'm not super familiar w/ DHO though, this might be impossible.

### CODE REVIEW

**Comment by bjoernwe** [comment]
> Added data-download instructions to README (lmk if I got it wrong)

Yes, that should be correct for the version you were working on. However, it turned out that the LFS were too large for Github's free-tier and I had to pay for it. So I'm going to remove all the LFS files again in the next PR.

**Comment by bjoernwe** [comment]
> added poetry.lock (I think that's typically checked in?)

I've read different opinions. But the canonical one is to check the file in, right? Let's try it.

**Comment by bjoernwe** [comment]
> I alway have problems in python where it can't load local libraries, e.g. when running python 1_crawl_messages.py. To get it working, I have to add sys.path.append("../"). Any ideas why?

That's not happening in PyCharm but when you run it from the command line, right? You can run the `python` command from the project's root like this: `python -m scraper.1_crawl_messages` (without `.py`).

**Comment by bjoernwe** [comment]
> It looks like author names have been anonymized. Any way to convert? (e.g. if I see an author on the site that I want to analyze here)

Yes, they are anonymized with a pipeline module in the scraper ([here](https://github.com/bjoernwe/dho-scrapy/blob/main/scraper/dho_scraper/pipelines.py#L31)). It's activated in [`settings.py:70`](https://github.com/bjoernwe/dho-scrapy/blob/main/scraper/dho_scraper/settings.py#L70). So you could disable the anonymization temporarily, if you like.

Anonymization is done with the [`codenamize`](https://github.com/jjmontesl/codenamize) library. So you could open an interactive shell and call `codenamize()` on the user name you are interested in.

**Comment by bjoernwe** [comment]
> It seems like message content might be missing newlines--known issue?

That's very possible, yes! I haven't missed the newlines yet. I was rather trying to get rid of as many special characters as possible. Would it be helpful to keep the newlines?

**Comment by bjoernwe** [comment]
> It would be nice to get just the main practice log for an author, rather than all their posts. I realize that's probably not super straightforward, but one heuristic might be "find the thread with the most posts by $AUTHOR, and filter for only their top-level posts." I'm not super familiar w/ DHO though, this might be impossible.

I'd say it's almost possible to get the practice logs instead of all messages. There's an example in [`examples/example_message_filtering.py`](https://github.com/bjoernwe/dho-scrapy/blob/main/examples/example_message_filtering.py#L33):

1) Filter messages for category `PracticeLogs`
1) Then filter threads for a certain author (i.e., who has started the thread). That gives you all the messages from a user's practice log threads, like [this one](https://www.dharmaoverground.org/discussion/-/message_boards/message/24869380).
1) You have mainly practice logs then but in between there may be messages from other users and responses to that. You can at least filter away the other users' posts (`filter_thread_responses(keep_op=True)` in the example).

It's not perfect because there can still responses from the practicing user to other people's comments. The optimal way to filter would be to through away all messages that are not top-level in the thread. But unfortunately, I gave up on parsing the the hierarchical message structure from the forum because it was annoyingly difficult to do so. If it turns out we need the hierarchical structure, we can find a way...

File: experiments/superb-owl/chatgpt/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Chat-GPT experiment
+
+To run this experiment:
+* set the `OPENAI_API_KEY` environment variable (you can use .envrc if you have direnv installed)
+* `python -m pip install openai`
```
Comment (bjoernwe) [inline experiments/superb-owl/chatgpt/README.md:5]: You can also add the dependency to the project-wide dependencies:
`poetry add openai --group experiments`

**Review by bjoernwe — APPROVED** [review]

**Comment by superb-owl** [comment]
> That's very possible, yes! I haven't missed the newlines yet. I was rather trying to get rid of as many special characters as possible. Would it be helpful to keep the newlines?

Nice to have more than anything--they don't carry much semantic content. I noticed it when displaying the current message in the UI I put together.

File: experiments/superb-owl/chatgpt/README.md (hunk 1 above)
Comment (superb-owl) [reply experiments/superb-owl/chatgpt/README.md:5]: Done!

**Comment by superb-owl** [comment]
Thanks for all the info! All that makes sense to me. Confirmed the `python -m` syntax works too 🎉

**Pull request merged**